Dockerfile

```diff
@@ -13,104 +13,146 @@
 # limitations under the License.
 
 ###############################################################################
-# Stage 1: Create the develop, test, and build environment
+# Stage 1: Create the developer image for the BUILDPLATFORM only
 ###############################################################################
-FROM  registry.access.redhat.com/ubi8/ubi-minimal:8.4 AS develop
+ARG GOLANG_VERSION=1.17
+FROM --platform=$BUILDPLATFORM registry.access.redhat.com/ubi8/go-toolset:$GOLANG_VERSION AS develop
 
-ARG GOLANG_VERSION=1.17.13
 ARG PROTOC_VERSION=21.5
 
 USER root
 
 # Install build and dev tools
-RUN microdnf install \
-    gcc \
-    gcc-c++ \
-    make \
-    vim \
-    findutils \
-    diffutils \
-    git \
-    wget \
-    tar \
-    unzip \
-    python3 \
-    nodejs && \
+RUN --mount=type=cache,target=/root/.cache/dnf:rw \
+    dnf install -y --nodocs --setopt=cachedir=/root/.cache/dnf \
+        python3 \
+        python3-pip \
+        nodejs \
+    && true
+
+# Install pre-commit
+ARG PIP_CACHE_DIR=/root/.cache/pip
+RUN --mount=type=cache,target=/root/.cache/pip \
```

Comment:
So I've been experimenting with this a bit. Does this match your understanding?
- the $HOME for `root` in `go-toolset` is not `/root/` so we need to set the `PIP_CACHE_DIR` env var to configure pip's cache directory
- `ARG` interpolation doesn't work in the `--mount`, so the value is copied instead of using `target=$PIP_CACHE_DIR`

If that's the case, I think it is slightly cleaner to set the PIP_CACHE_DIR env var in the `RUN` command since the `ARG` is not really configurable.

Though I also found https://github.com/moby/buildkit/issues/815 and it sounds like using ARGs in `--mount` can now work, but requires a newer Dockerfile syntax `# syntax=docker/dockerfile:1.3.0`.

Reply:
Hi 👋

Unless the host system (github actions) would build python wheels from source, there is no speedup from mounting pip's cache dir. We generally recommend to [set](https://github.com/alpha-affinity/snakepacker/blob/d1d7a7996fed2d4106e2a567404612b2bfda6340/base/Dockerfile#L24) PIP_NO_CACHE_DIR and rely on docker layer caching by running pip install before copying over source code. That way, you reuse the registry's layer as long as your `requirements.txt` doesn't change. Here's a live multi-arch [example](https://github.com/alpha-affinity/libpostal-fastapi/blob/c200764b9aa8d015ff9d1d5015f42296e238f600/Dockerfile#L21) that combines a compiled C binary (libpostal) with python builds that use it.

Reply:
Good comments/questions @tjohnson31415 @ddelange !

> Unless the host system (github actions) would build python wheels from source, there is no speedup from mounting pip's cache dir.

The cache mount is useful for subsequent builds, multi-stage and multi-platform builds. I have seen it being used and speedup the GitHub action builds quite a bit, with significant time savings for installing `grpc` and `tensorflow`

> the $HOME for `root` in `go-toolset` is not `/root/` so we need to set the `PIP_CACHE_DIR` env var to configure pip's cache directory

Yes, otherwise we see this [error](https://github.com/kserve/modelmesh-runtime-adapter/actions/runs/4675184385/jobs/8280043415#step:7:492):
```
#13 12.99 The directory '/opt/app-root/src/.cache/pip/http' or its parent directory is not owned by the current user and the cache has been disabled. Please check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
#13 13.00 The directory '/opt/app-root/src/.cache/pip' or its parent directory is not owned by the current user and caching wheels has been disabled. check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
```

However, for `pip`, the inline `--cache-dir` CLI flag did not set the cache correctly for both HTTP download and the Wheel cache directories. So I use the `ARG PIP_CACHE_DIR=/root/.cache/pip` which works for both.

> `ARG` interpolation doesn't work in the `--mount`, so the value is copied instead of using `target=$PIP_CACHE_DIR`

Yes. I have not tried the new syntax `# syntax=docker/dockerfile:1.3.0` though. That's a good find.

---

References:
- https://www.docker.com/blog/faster-multi-platform-builds-dockerfile-cross-compilation-guide/
- https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/reference.md#run---mounttypecache
- https://docs.docker.com/build/cache/
- https://github.com/moby/buildkit/issues/1673